libpod/container_top.go

```diff
@@ -0,0 +1,177 @@
+package libpod
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"encoding/json"
+	"github.com/pkg/errors"
+	"github.com/projectatomic/libpod/utils"
+	"github.com/sirupsen/logrus"
+)
+
+// This can be removed is stats is merged
+// CGroupParent - should be used from stats
+var CGroupParent = "/libpod_parent"
+
+// GetContainerPids reads sysfs to obtain the pids associated with the container's cgroup
+func (c *Container) GetContainerPids() ([]string, error) {
+	taskFile := filepath.Join("/sys/fs/cgroup/pids", CGroupParent, fmt.Sprintf("libpod-conmon-%s", c.ID()), c.ID(), "tasks")
+	logrus.Debug("reading pids from ", taskFile)
+	content, err := ioutil.ReadFile(taskFile)
+	if err != nil {
+		return []string{}, errors.Wrapf(err, "unable to read pids from %s", taskFile)
+	}
+	return strings.Split(string(content), "\n"), nil
+}
+
+// GetContainerPidInformation calls ps with the appropriate options and returns
+// the results as a string
+func (c *Container) GetContainerPidInformation(args []string) ([]string, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if err := c.syncContainer(); err != nil {
+		return []string{}, errors.Wrapf(err, "error updating container %s state", c.ID())
+	}
+	pids, err := c.GetContainerPids()
+	if err != nil {
+		return []string{}, errors.Wrapf(err, "unable to obtain pids for ", c.ID())
+	}
+	args = append(args, "-p", JoinNoTrailingSep(pids, ","))
+	logrus.Debug("Executing: ", strings.Join(args, " "))
+	results, err := utils.ExecCmd("ps", args...)
+	if err != nil {
+		return []string{}, errors.Wrapf(err, "unable to obtain information about pids")
+	}
+	return strings.Split(results, "\n"), nil
+}
+
+//JoinNoTrailingSep takes a string array and separator and uses
+// strings.Join to combine then,  but then extracts the trailing
+// separator from the string.
+func JoinNoTrailingSep(s []string, sep string) string {
+	newString := strings.Join(s, sep)
+	lastIndex := strings.LastIndex(newString, sep)
+	return newString[0:lastIndex]
+}
+
+// PSDataToParams converts a string array of data and its headers to an
+// arra if PSParams
+func PSDataToPSParams(data []string, headers []string) ([]PSParams, error) {
```

Comment:
Can probably be private

Reply:
Nevermind, wasn't looking at how this was used

Reply:
Since this is only ever used with the output of GetContainerPidInformation, maybe make this GetContainerPidInformationJSON() or something and call GetContainerPidInformation in here, then do the conversion